GPU/GPUbenchmark/cuda/Kernels.cu

```diff
@@ -130,14 +130,50 @@ __global__ void copy_k(
 template <class chunk_t>
 __global__ void rand_read_k(
   chunk_t* chunkPtr,
-  size_t chunkSize)
+  size_t chunkSize,
+  int prime)
 {
   chunk_t sink{0};
-  BSDRnd r{};
-  for (size_t i = threadIdx.x; i < chunkSize; i += blockDim.x) {
-    sink = chunkPtr[i];
+  for (size_t i = blockIdx.x * blockDim.x + threadIdx.x; i < chunkSize; i += blockDim.x * gridDim.x) {
+    sink += chunkPtr[(i * prime) % chunkSize];
+  }
+  chunkPtr[threadIdx.x] = sink;
+}
+
+// Random write
+template <class chunk_t>
+__global__ void rand_write_k(
+  chunk_t* chunkPtr,
+  size_t chunkSize,
+  int prime)
+{
+  for (size_t i = blockIdx.x * blockDim.x + threadIdx.x; i < chunkSize; i += blockDim.x * gridDim.x) {
+    chunkPtr[(i * prime) % chunkSize] = 0;
+  }
+}
+
+template <>
+__global__ void rand_write_k(
+  int4* chunkPtr,
+  size_t chunkSize,
+  int prime)
+{
+  for (size_t i = blockIdx.x * blockDim.x + threadIdx.x; i < chunkSize; i += blockDim.x * gridDim.x) {
+    chunkPtr[(i * prime) % chunkSize] = {0, 1, 0, 0};
+  };
+}
+
+// Random copy
+template <class chunk_t>
+__global__ void rand_copy_k(
+  chunk_t* chunkPtr,
+  size_t chunkSize,
+  int prime)
+{
+  size_t offset = chunkSize / 2;
+  for (size_t i = blockIdx.x * blockDim.x + threadIdx.x; i < offset; i += blockDim.x * gridDim.x) {
+    chunkPtr[i] = chunkPtr[(offset + i * prime) % offset]; // might be % = 0...
```

Comment:
you should apply the randomization also to the destination, i.e. `chunkPtr[(i * prime) % offset]`
And for the source part, the offset should not be taken into the modulo:
`chunkPtr[offset + (i * primt) % offset]`